src/engines/basic-render-engine.ts

```diff
@@ -182,6 +203,14 @@ export class BasicRenderEngine extends EventEmitter {
         this.ctx.fillRect(x, y, w, this.blockHeight);
     }
 
+    renderIntervalBlock(_color: string, x: number, y: number, w: number): void {
```

Comment:
This operation should use the same approach as for other ctx state changes. And you can't use the same queue for render intervals as for rectangles. It should keep the `fillStyle` and change the ctx state only if necessary. And yes, all operations should also be queued and sorted by their fillstyle and color in order to change the ctx properties as few times as possible. Otherwise, more than 500-1000 objects will already have performance problems.(
https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API/Tutorial/Optimizing_canvas#more_tips
And it will be nice if you read this post about canvas performance.) All these tips are worth using when developing because we work with a large number of objects.